File: Game-1-modular/systems/class_system.py
"""Class system for managing character class selection and bonuses"""

import json
from typing import Optional, Dict, Callable, List

from data.models import ClassDefinition


# ── §15 trap 7 — JSON-driven tag→tool bonus table ────────────────────
#
# Loaded lazily on first use. Falls back to the historical Python
# constants if classes-1.JSON is missing or doesn't carry a
# ``metadata.tagToolBonuses`` block.

_FALLBACK_TAG_TOOL_BONUSES: Dict[str, Dict[str, float]] = {
    'axe': {'nature': 0.10, 'gathering': 0.05},
    'pickaxe': {'gathering': 0.10, 'explorer': 0.05},
    'toolDamage': {'physical': 0.05, 'melee': 0.05},
}

_TAG_TOOL_BONUSES_CACHE: Optional[Dict[str, Dict[str, float]]] = None


def _load_tag_tool_bonuses() -> Dict[str, Dict[str, float]]:
    global _TAG_TOOL_BONUSES_CACHE
    if _TAG_TOOL_BONUSES_CACHE is not None:
        return _TAG_TOOL_BONUSES_CACHE
    try:
        from core.paths import get_resource_path
        path = get_resource_path("progression/classes-1.JSON")
        if not path.exists():
            _TAG_TOOL_BONUSES_CACHE = _FALLBACK_TAG_TOOL_BONUSES
            return _TAG_TOOL_BONUSES_CACHE
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
        metadata = data.get("metadata", {}) or {}
        table = metadata.get("tagToolBonuses") or {}
        # Drop the editor-only _note key if present
        cleaned = {k: {tk: float(tv) for tk, tv in v.items() if not tk.startswith("_")}
                   for k, v in table.items() if not k.startswith("_")}
        _TAG_TOOL_BONUSES_CACHE = cleaned if cleaned else _FALLBACK_TAG_TOOL_BONUSES
    except Exception:
        _TAG_TOOL_BONUSES_CACHE = _FALLBACK_TAG_TOOL_BONUSES
    return _TAG_TOOL_BONUSES_CACHE


def reload_tag_tool_bonuses() -> None:
    """Clear the cache so the next call re-reads the JSON."""
    global _TAG_TOOL_BONUSES_CACHE
    _TAG_TOOL_BONUSES_CACHE = None


class ClassSystem:
    def __init__(self):
        self.current_class: Optional[ClassDefinition] = None
        self._on_class_set_callbacks: List[Callable[[ClassDefinition], None]] = []

    def set_class(self, class_def: ClassDefinition):
        self.current_class = class_def
        # Publish CLASS_CHANGED to GameEventBus for World Memory System
        try:
            from events.event_bus import get_event_bus
            get_event_bus().publish("CLASS_CHANGED", {
                "actor_id": "player",
                "class_id": class_def.class_id,
            })
        except Exception:
            pass
        # Notify any registered callbacks
        for callback in self._on_class_set_callbacks:
            callback(class_def)

    def register_on_class_set(self, callback: Callable[[ClassDefinition], None]):
        """Register a callback to be called when a class is set."""
        self._on_class_set_callbacks.append(callback)

    def get_bonus(self, bonus_type: str) -> float:
        if not self.current_class:
            return 0.0
        return self.current_class.bonuses.get(bonus_type, 0.0)

    def get_tool_efficiency_bonus(self, tool_type: str) -> float:
        """Get tool efficiency bonus based on class tags.

        §15 trap 7: bonuses come from
        ``progression/classes-1.JSON > metadata.tagToolBonuses``.
        """
        if not self.current_class or not self.current_class.tags:
            return 0.0
        tags = set(t.lower() for t in self.current_class.tags)
        per_tool = _load_tag_tool_bonuses().get(tool_type, {})
        return sum(v for k, v in per_tool.items() if k in tags)

    def get_tool_damage_bonus(self) -> float:
        """Get tool damage bonus for combat use based on class tags."""
        if not self.current_class or not self.current_class.tags:
            return 0.0
        tags = set(t.lower() for t in self.current_class.tags)
        per_tool = _load_tag_tool_bonuses().get("toolDamage", {})
        return sum(v for k, v in per_tool.items() if k in tags)

File: Unity/Assets/Scripts/Game1.Systems/Progression/ClassSystem.cs
// ============================================================================
// Game1.Systems.Progression.ClassSystem
// Migrated from: systems/class_system.py (70 lines), data/models/classes.py (47 lines)
// Migration phase: 4
// Date: 2026-02-13
//
// Character class system with tag-driven bonuses.
// 6 classes, each with tags determining skill affinity, tool bonuses, and identity.
// Max class affinity bonus: 20% (5% per matching tag).
// ============================================================================

using System;
using System.Collections.Generic;
using System.Linq;
using Game1.Core;

namespace Game1.Systems.Progression
{
    // ========================================================================
    // Class Definition (data model)
    // ========================================================================

    /// <summary>
    /// Definition for a character class with tag-driven identity.
    /// Loaded from classes-1.JSON.
    /// Matches Python ClassDefinition dataclass exactly.
    /// </summary>
    public class ClassDefinition
    {
        public string ClassId { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public Dictionary<string, float> Bonuses { get; set; } = new();
        public string StartingSkill { get; set; } = "";
        public List<string> RecommendedStats { get; set; } = new();
        public List<string> Tags { get; set; } = new();
        public List<string> PreferredDamageTypes { get; set; } = new();
        public string PreferredArmorType { get; set; } = "";

        /// <summary>
        /// Check if class has a specific tag (case-insensitive).
        /// Matches Python ClassDefinition.has_tag().
        /// </summary>
        public bool HasTag(string tag)
        {
            return Tags.Any(t => t.Equals(tag, StringComparison.OrdinalIgnoreCase));
        }

        /// <summary>
        /// Calculate skill affinity bonus based on tag overlap.
        /// Each matching tag adds 5% bonus, up to 20% max.
        /// Matches Python ClassDefinition.get_skill_affinity_bonus() exactly.
        /// </summary>
        public float GetSkillAffinityBonus(List<string> skillTags)
        {
            if (skillTags == null || skillTags.Count == 0 || Tags.Count == 0)
                return 0f;

            var classTagsLower = new HashSet<string>(Tags.Select(t => t.ToLowerInvariant()));
            var skillTagsLower = new HashSet<string>(skillTags.Select(t => t.ToLowerInvariant()));

            int matchCount = classTagsLower.Intersect(skillTagsLower).Count();

            const float bonusPerTag = 0.05f;
            const float maxBonus = GameConfig.MaxClassAffinityBonus; // 0.20

            return MathF.Min(matchCount * bonusPerTag, maxBonus);
        }
    }

    // ========================================================================
    // Class System
    // ========================================================================

    /// <summary>
    /// Manages character class selection and provides tag-driven bonuses.
    /// Matches Python ClassSystem class exactly.
    /// </summary>
    public class ClassSystem
    {
        /// <summary>Currently selected class (null before class selection).</summary>
        public ClassDefinition CurrentClass { get; private set; }

        /// <summary>
        /// Reference to class database. Set after database singleton is initialized.
        /// </summary>
        public Dictionary<string, ClassDefinition> ClassDatabase { get; set; } = new();

        // Callbacks for class change events
        private readonly List<Action<ClassDefinition>> _onClassSetCallbacks = new();

        // ====================================================================
        // Class Selection
        // ====================================================================

        /// <summary>
        /// Select a class by ID from the database.
        /// </summary>
        public bool SelectClass(string classId)
        {
            if (ClassDatabase.TryGetValue(classId, out var classDef))
            {
                SetClass(classDef);
                return true;
            }
            return false;
        }

        /// <summary>
        /// Set the current class directly.
        /// Matches Python ClassSystem.set_class().
        /// </summary>
        public void SetClass(ClassDefinition classDef)
        {
            CurrentClass = classDef;
            foreach (var callback in _onClassSetCallbacks)
                callback(classDef);
        }

        /// <summary>
        /// Register a callback for class changes (e.g., to recalculate stats).
        /// Matches Python ClassSystem.register_on_class_set().
        /// </summary>
        public void RegisterOnClassSet(Action<ClassDefinition> callback)
        {
            _onClassSetCallbacks.Add(callback);
        }

        // ====================================================================
        // Bonus Queries
        // ====================================================================

        /// <summary>
        /// Get a specific bonus value from the current class.
        /// Matches Python ClassSystem.get_bonus().
        /// </summary>
        public float GetBonus(string bonusType)
        {
            if (CurrentClass == null) return 0f;
            return CurrentClass.Bonuses.TryGetValue(bonusType, out float val) ? val : 0f;
        }

        /// <summary>
        /// Get all class bonuses as a dictionary.
        /// </summary>
        public Dictionary<string, float> GetClassBonuses()
        {
            if (CurrentClass == null) return new Dictionary<string, float>();
            return new Dictionary<string, float>(CurrentClass.Bonuses);
        }

        /// <summary>
        /// Get skill affinity bonus for the current class based on skill tags.
        /// Each matching tag adds 5% bonus, up to 20% max.
        /// </summary>
        public float GetClassAffinity(List<string> skillTags)
        {
            return CurrentClass?.GetSkillAffinityBonus(skillTags) ?? 0f;
        }

        /// <summary>
        /// Get discipline affinity bonus. Maps discipline name to class bonus keys.
        /// </summary>
        public float GetDisciplineAffinity(string discipline)
        {
            if (CurrentClass == null) return 0f;

            string key = $"{discipline.ToLowerInvariant()}_bonus";
            return CurrentClass.Bonuses.TryGetValue(key, out float val) ? val : 0f;
        }

        /// <summary>
        /// Get tool efficiency bonus based on class tags.
        /// Matches Python ClassSystem.get_tool_efficiency_bonus() exactly.
        ///
        /// Tag-driven bonuses:
        ///   'nature' or 'gathering': +10% axe efficiency
        ///   'gathering' or 'explorer': +10% pickaxe efficiency
        ///   'physical' or 'melee': +5% tool damage (via get_tool_damage_bonus)
        /// </summary>
        public float GetToolEfficiencyBonus(string toolType)
        {
            if (CurrentClass?.Tags == null || CurrentClass.Tags.Count == 0)
                return 0f;

            var tags = new HashSet<string>(CurrentClass.Tags.Select(t => t.ToLowerInvariant()));
            float bonus = 0f;

            switch (toolType.ToLowerInvariant())
            {
                case "axe":
                    if (tags.Contains("nature")) bonus += 0.10f;
                    if (tags.Contains("gathering")) bonus += 0.05f;
                    break;
                case "pickaxe":
                    if (tags.Contains("gathering")) bonus += 0.10f;
                    if (tags.Contains("explorer")) bonus += 0.05f;
                    break;
            }

            return bonus;
        }

        /// <summary>
        /// Get tool damage bonus for combat use based on class tags.
        /// Matches Python ClassSystem.get_tool_damage_bonus() exactly.
        /// </summary>
        public float GetToolDamageBonus()
        {
            if (CurrentClass?.Tags == null || CurrentClass.Tags.Count == 0)
                return 0f;

            var tags = new HashSet<string>(CurrentClass.Tags.Select(t => t.ToLowerInvariant()));
            float bonus = 0f;

            if (tags.Contains("physical")) bonus += 0.05f;
            if (tags.Contains("melee")) bonus += 0.05f;

            return bonus;
        }

        // ====================================================================
        // Save/Load
        // ====================================================================

        /// <summary>
        /// Get current class ID for save data.
        /// </summary>
        public string GetCurrentClassId()
        {
            return CurrentClass?.ClassId;
        }

        /// <summary>
        /// Restore class from save data by class ID.
        /// </summary>
        public void RestoreFromSave(string classId)
        {
            if (!string.IsNullOrEmpty(classId) &&
                ClassDatabase.TryGetValue(classId, out var classDef))
            {
                CurrentClass = classDef;
                // Do NOT fire callbacks during restore (avoid side effects)
            }
        }
    }
}
